glibc: 2.27 -> 2.30 WIP

lblasc opened the pull request:
###### Motivation for this change

Bump to latest glibc.

Started to fix core packages, managed to build and test `rustc` on x86_64.

- can someone please point me to latest bootstrap files for i686 and x86_64 on tarballs.nixos.org so I can remove my private mirror from `pkgs/stdenv/linux/bootstrap-files/*` ? @grahamc @Mic92
- hydra jobset for this would be great
- test darwin

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @edolstra

lblasc committed:
cpython: fetch darwin-libutil.patch to nixpkgs

python3 is now required buildInput for glibc>=2.29.
Remove fetchpatch to solve infinite recursion in
glibc bootstrap process.

lblasc committed:
glibc: 2.27 -> 2.30

lblasc committed:
gnulib: 20190326 -> 20190811

lblasc committed:
gnum4: fix build with glibc>=2.28

lblasc committed:
sharutils: fix build with glibc>=2.28

lblasc committed:
bootstrap-files: bump WIP

jonringer commented:
You're a madman, and i support you on your journey :heart: 

However, such a universal package may cause a lot more heart-ache than it's worth. (don't want you to get burnt out)

Mic92 commented on pkgs/development/libraries/glibc/common.nix:
```diff
@@ -138,7 +139,7 @@ stdenv.mkDerivation ({
   depsBuildBuild = [ buildPackages.stdenv.cc ];
   nativeBuildInputs = [ bison ];
-  buildInputs = [ linuxHeaders ] ++ lib.optionals withGd [ gd libpng ];
+  buildInputs = [ linuxHeaders python3 ] ++ lib.optionals withGd [ gd libpng ];
```
Is the interpreter used during the build? If so, you want to move it to `nativeBuildInputs` instead.

Mic92 commented:
cc @LnL7

FRidh commented:
These kind of bumps typically takes months to get ready, so be prepared. It's good to get such an upgrade in, however, I just want to point also the openssl bump which could really use some help
https://github.com/NixOS/nixpkgs/pull/22357

LnL7 commented on pkgs/stdenv/linux/bootstrap-files/x86_64.nix:
```diff
@@ -1,9 +1,12 @@
+  };
+
   bootstrapTools = import <nix/fetchurl.nix> {
-    url = http://tarballs.nixos.org/stdenv-linux/x86_64/4907fc9e8d0d82b28b3c56e3a478a2882f1d700f/bootstrap-tools.tar.xz;
```
Why do these need to be updated? While we have to do this regularly for darwin compatibility, there hasn't been a reason to update these for linux since early 2016.

lblasc replied:
Minimal gcc version for building glibc is now 6.2, bootstrap tools provide only version 5. Also some other libs are out of date.

vcunat replied:
Hmm, that _could_ be worked around by reorganizing the bootstrapping process, but I'd probably choose the easier route of updating the bootstrap-tools.

LnL7 commented on pkgs/development/interpreters/python/cpython/3.7/darwin-libutil.patch:
```diff
@@ -0,0 +1,23 @@
+diff --git a/Modules/posixmodule.c b/Modules/posixmodule.c
```
How is this related to the rest of the changes?

lblasc replied:
python3 is now required for building glibc. When python is built in bootstrap phase fetchpatch causes infinite recursion. I've solved this by including patch in nixpkgs source. Other ideas are welcomed.

FRidh replied:
in this case https://github.com/NixOS/nixpkgs/pull/66762 may come in handy

vcunat replied:
Well, run-time closure reduction isn't important there, but reducing the build-time closure could speed up bootstrapping (though that by itself doesn't sound a big deal to me).

LnL7 reviewed (commented):
I'm not particularly familiar with linux bootstrapping so I can't give much input on this. Since glibc isn't used on darwin only the updates/patches to common unix tools like findutils will have impact there.

vcunat commented:
Do you know if the locale archive format has changed in an incompatible way?  If not, I would **not** rename the `LOCALE_ARCHIVE_2_27` name.

vcunat committed:
stdenv bootstrap: disable dejagnu tests

There's a failure - I don't know why.  The checks from "normal" dejagnu
still succeed, so I assume this commit is quite a safe work-around.

vcunat commented:
^^ I'm not sure why dejagnu tests didn't cause a problem for you, but I don't expect we don't need to track this thing down (and can keep the commit).

vcunat commented:
-
  > can someone please point me to latest bootstrap files for i686 and x86_64 on tarballs.nixos.org so I can remove my private mirror from `pkgs/stdenv/linux/bootstrap-files/*` ?

  We're surely using the latest uploaded version already.  This upload process is manual.

- 2.30 was released a couple weeks ago, so I'd probably avoid putting that into 19.09 (to be forked from nixpkgs master in 2-3 weeks).  One option might be to go with 2.29 (e.g. + stable patches) for now, which would also avoid the need to update the bootstrap tools at this point (I think).  But I don't expect we can catch that anyway, given the IDN issue below.

#### [IDN](https://en.wikipedia.org/wiki/Internationalized_domain_name)

This DNS stuff is actually why I got stuck with 2.28 update and didn't finish it.  As the PR is now, executables where `dlopen()` can't find libidn2 regress (since 2.27) when processing non-ASCII domain names by the standard libc calls like `getaddrinfo()`.  I believe most people would be unaffected, but in some countries IDN names are very common (correlated with not using latin-based alphabet in their main languages).

  My *plan* from about a year ago :-) was... to always link libidn2 when linking libc.  The added closure (wrt. glib) isn't too huge (<2 MB) and the libraries are already present in closures of most systems anyway.  The technical approach was to modify the `libc.so` linker script by adding libidn2.  Of course, that would create a cycle, so the plan was to complicate bootstrapping a bit where the final glibc package would be a thin wrapper just extending this linker script.

  Now... the question is whether this dance is worth fixing this regression, or perhaps if you can suggest some better way of fixing it.

vcunat committed:
glibc: depend on libidn2 (and libunistring, transitively)

It's a bit hacky, but ATM I can't see any better way for glibc >= 2.28.

Signed-off-by: Luka Blaskovic <lblasc@tvbeat.com>

vcunat commented:
So, I focused on the IDN stuff today and came with a set of hacks that seems to work just fine: 9cd4b636.  I'm not so sure of that, so I didn't push it into this PR (yet).

After we get a bit more resolution about what to do, I can create a Hydra jobset.  As it is now, I think glibc-2.29 for NixOS 19.09 sounds realistic as well to me (/cc managers: @lheckemann, @disassembler), but we can still just wait until after the fork and go straight for 2.30.

lblasc commented:
@vcunat great work, thanks for looking into this.

* reverted back to LOCALE_ARCHIVE_2_27
* about dejagnu :) I forgot to mention that I have locally disabled some tests because I'm building in LXC container with apparmor enabled and this causes some tests to fail (like e2fsprogs), seems like dejagnu failing tests are not related to apparmor.
* I like the IDN approach you taken, I'm not the expert in NixOS bootstrapping but this patch looks ok to me.

About next steps, my suggestion is that we wait for fork and then go straight for 2.30. Until then we can run separate hydra jobset and fix things, I'm sure stuff will brake :)

vcunat commented:
Hmm, I forgot about `libunistring` tests &ndash; those get broken, and we use that version as the final library.  Hopefully that can be fixed easily somehow.

For reference, I really verified that glibc IDN works in master, gets broken here and fixed by my patch.  Simple test is to check if `${glibc.bin}/bin/getent ahosts háčkyčárky.cz` gives empty output or something like
```
2001:1488:0:3::5 STREAM háčkyčárky.cz
2001:1488:0:3::5 DGRAM  
2001:1488:0:3::5 RAW    
217.31.205.51   STREAM 
217.31.205.51   DGRAM  
217.31.205.51   RAW    
```
_You may note that SW very often doesn't rely on glibc support, usually by using libidn* directly (even our `ping` command), so the issue probably wouldn't be a big deal._

lblasc committed:
libunistring: disable tests

lblasc commented:
Spent few hours debugging `libunistring` tests, for now no luck, seems like failures are related to gnulib version used in this lib. I've disabled tests for now.

Did rebase, also added latest work from #66748.

vcunat commented:
One thing we need to keep in mind is that there are multiple platforms (using glibc).  I did some investigation in [a temporary branch](https://github.com/vcunat/nixpkgs/commits/p/update-stdenv-bootstrap-tools):
- x86_64: tested /nix/store/21zzaia2nca8p5vpcrklrdqlj843q2wf-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
- x86_64-musl: seems to be 7.3 which should be OK; also I don't expect it uses glibc anyway
- i686: tested /nix/store/l1r0lyzdr2n5qqz68rwxnh71qjd5s4ay-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
- aarch64: /nix/store/fwrp466vixrzc63l4kvlslalfsxkjnvk-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
	and optionally with /nix/store/fwrp466vixrzc63l4kvlslalfsxkjnvk-stdenv-bootstrap-tools/on-server/busybox
	* FIXME: when combined with the glibc-2.30 branch, bzip2 won't build during bootstrapping (<- python3 <- glibc)
		/nix/store/w9yzgh5sr7i65icbb3gcdjyc7y1zksfc-bzip2-1.0.6.0.1.drv
- aarch64-musl, armv6l-musl: seems to be 6.4 which should be OK; also I don't expect it uses glibc anyway

- loongson2f: probably didn't work anymore, not changed hash since 2010 (40405d03ac)
- armv7l, armv6l, armv5tel: I see gcc-5.4; @dezgeg: do you have some reasonable way of testing these?  I have a v7 router, but it perhaps won't be easy for me to set up nix in there, even though LXC is supported.

vcunat commented:
I should note that aarch64 is a blocker;  I used our [community machine](https://github.com/nix-community/aarch64-build-box) for testing.  After we manage to fix at least basic bootstrap in there, I can bother other people to upload all these tarballs.  The other platforms (like 32-bit ARMs) are not that critical and could probably be left... to maintainers of those platforms.

tollb commented:
It appears that the locale changes that come with a glibc upgrade from 2.27 to 2.28 or higher may impact PostgreSQL users with existing databases. The updated locales affect sort ordering and that can cause incorrect behavior. See: [https://wiki.postgresql.org/wiki/Locale_data_changes](url).
@thoughtpolice @ocharles @danbst @globin

danbst commented:
@tollb thanks for the notice. They write that Ubuntu had gone with "break-it" path, so should we. Link to that wiki page in release doc is enough.

Ma27 committed:
glibc: remove outdated patches

Co-authored-by: Luka Blaskovic <lblasc@znode.net>

Ma27 committed:
bzip2: fix aarch64 build

Ma27 committed:
expect: fix aarch64 build

Ma27 committed:
stdenv: rebuild aarch64 bootstrap files with gcc8

The old bootstrapping files would break with glibc 2.30 as at least gcc6
is needed.

Ma27 commented:
fpletz and I decided that we'd try to finish this. The changes above were needed to build `pkgs.hello` on the aarch64 community builder. We're creating a Hydra jobset now.

fpletz commented:
Hydra jobset is here: https://hydra.nixos.org/jobset/nixpkgs/glibc-230

lblasc commented:
@Ma27 thx for aarch64 patches and the effort. @vcunat I've tried to use bootstrap-tools (x86_64 and i686) from your [temporary branch](https://github.com/vcunat/nixpkgs/commits/p/update-stdenv-bootstrap-tools) seems like those were only available for testing, can you please upload new ones (now with gcc9).

vcunat commented:
They were never uploaded.  I just prepared the expressions, but I don't even have credentials to upload myself, and I was waiting for the aarch64 blocker anyway.  The `/nix/store/` paths above should still be in the official cache (so you can fetch them by `nix build /nix/store/foo`), but after that I used `nix-prefetch-url file:///nix/store/foo/...` or something similar to get the individual files (I don't remember details anymore).

lblasc commented:
OK, I've update bootstrap files for x86_64 and i686 from the latest master, also cherry-picked one TimeDate patch which is staging-next (but not in staging) will remove it when it will be merged into staging.

lblasc committed:
libffi: only run tests on musl targets

Python3 is now needed for for glibc build which pulls libffi,
so I suppose libffi is now part of stdenv on linux with glibc
as on darwin.

lblasc commented:
Found one more blocker, libffi, seems like it is now stdenv dependency (python3 is pulling it), so tests are now enabled only on musl targets.

@fpletz please reevaluate hydra jobset.

Ma27 commented:
Started a new evaluation :)

lblasc committed:
ppp: fix build with glibc>=2.28

Ma27 commented:
Decided to merge staging into our branch again. It seems as staging had some issues during my last merge related to the partial merge of the structured-attrs feature that seem to be fixed now. I also started a new evaluation of the Hydra jobset (btw if you have a Google-Account, you should be able to login with that one on Hydra to trigger jobset evals, sorry forgot to mention that earlier :/)

falsifian commented:
I think I was added because I'm listed as the ppp maintainer. Unfortunately I don't currently use ppp so am not really set up to test the change and should probably remove myself as maintainer. Sorry, I should probably have done this earlier.

What's the best thing to do in this case? Should I post on Discourse asking for a new maintainer, and if nobody shows up, remove ppp from nixpkgs?

vcunat commented:
I just removed you in 9e0c8f0cbed.  Of course it will be nice if someone steps up, but I don't think that's sufficient reason to (immediately) remove a package.  After all, we have quite a large fraction without any maintainer...

Ma27 commented:
Just fixed `libselinux` on staging (780fb563c74171aafc2fb802cfdebd56a68d0705) which seems to have caused most of the new build failures on our latest eval :)

jonringer commented:
ironically, the python3 bump was reverted https://github.com/NixOS/nixpkgs/pull/77047/commits/eb1369670b5a4e616ff0cf4100616479b1fa3064

lblasc commented:
There were some conflicts in merge with latest `staging`, so I did rebase to skip conflict resolution in merge commit. Only thing I modified is `glibc: remove outdated patches` commit with removal of some new patches which are not needed in 2.30.
New evaluation is started :)
I will continue working on `syslinux`, which is another blocker.

lblasc committed:
jfsutils: fix build with glibc>=2.28

lblasc committed:
oathToolkit: fix build with glibc>=2.28

Ma27 committed:
postfix: fix build w/glibc 2.30

Ma27 commented:
Just realized that we probably need to remove old gcc versions (4.x, 5) and llvm (3.x) versions as they break with glibc>=2.28, already started working on this, afaics most affected pkgs need an update to build with a newer compiler.

Ma27 commented:
The last batch of commits aims to remove gcc5 as this relies on headers that were removed in glibc 2.28 (like `<sys/ustat.h>`). Not sure what others think of this, but I guess that some of those can be cherry-picked directly to master to reduce the diff here :)

lblasc commented:
@Ma27 great work, can you open separate pull request for old gcc removal? I think it is better to merge those cleanup patches separately, also include removal of gcc49 (should be easy to remove).

Ma27 commented:
> great work, can you open separate pull request for old gcc removal?

:+1: 

> also include removal of gcc49

currently working on that one :)

Ma27 committed:
expect: remove hack for aarch64 build

Not sure if I missed something or the issue got fixed later, but it's
not needed anymore to pass a `--build` flag to the configure script on
aarch64.

The autoreconfHooks are still needed for expect and bzip2.

Ma27 committed:
seasocks: fix build w/glibc-2.30

Ma27 committed:
xnbd: fix build w/glibc-2.30

Ma27 committed:
rr: 5.2.0 -> 5.3.0, fix build w/glibc-2.30

Ma27 committed:
diod: fix build w/glibc-2.30

For compatibility with newer glibc versions, `sys/sysmacros.h` needs to
be included.

Ma27 committed:
distrho: fix build w/glibc-2.30

Ma27 committed:
gcal: fix build w/glibc-2.30

See https://lists.gnu.org/archive/html/bug-gcal/2018-10/msg00000.html
for further reference.

Ma27 committed:
drbd: fix build w/glibc-2.30

Decided to not bump to a later version as they decided to split the repo
which would've made change way more complicated.

Ma27 committed:
xen_4_8: fix build w/glibc-2.30

The header (and the `sysctl`-function) were unused, so the inclusion can be
removed safely.

This was done by upstream in 4.9 as well[1].

[1] https://github.com/xen-project/xen/commit/74858c9eaf335658b4c19cb4a848bb721065c233

Ma27 commented:
As suggested, I removed the syslinux fix as it's already fixed (and even updated) in a different PR. Also removed the gcc49/5 stuff here to reduce the diff a little bit :)

Ma27 committed:
mono: fix build w/glibc-2.30

Ma27 committed:
foundationdb61: fix build w/glibc-2.30

Ma27 committed:
spaceFM: fix build w/glibc-2.30

Ma27 committed:
xen_4_8-{light,slim}: fix build

Ma27 committed:
Merge branch 'staging' into glibc230

Ma27 commented:
Merged recent staging and triggered a new eval :)

Ma27 committed:
mono4: fix build w/glibc-2.30

Ma27 committed:
ext4magic: fix build w/glibc-2.30

Ma27 committed:
vit: fix tests

Ma27 committed:
intel-gpu-tools: 1.23 -> 1.24, fix build w/glibc-2.30

Ma27 committed:
jfbview: fix build w/glibc-2.30

Ma27 committed:
udftools: fix build w/glibc-2.30

Ma27 committed:
wcc: fix build w/glibc-2.30

Ma27 committed:
coz: 0.2.0 -> 0.2.1, fix build w/glibc-2.30

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/nixos-20-03-feature-freeze/5655/4

Ma27 committed:
z88dk-unstable: 2019-05-09 -> 2020-01-27, fix build w/glibc-2.30

Ma27 committed:
ttyrec: mark as broken

Requires svr4-specific stuff to work around "out of ptys"-error[1],
however this API has been dropped in glibc 2.30 and as the package is
fairly old, there doesn't seem to be an actual fix available.

[1] https://github.com/mjording/ttyrec/pull/2

Ma27 committed:
tiscamera: 0.9.1 -> 0.11.1, fix build w/glibc-2.30

Ma27 committed:
Merge branch 'staging' into glibc230

Conflicts:
	pkgs/applications/misc/vit/default.nix

Ma27 committed:
eresi: fix build w/glibc-2.30

Ma27 committed:
distrho: remove glibc patch

Package was updated on staging and contains the patch in its sources
now.

Ma27 committed:
gftp: fix build w/glibc-2.30

Ma27 committed:
ssw: init at 0.3

GNU spread-sheet widget. Needed for pspp-1.2.0.

Ma27 committed:
pspp: 1.0.1 -> 1.2.0, fix build w/glibc-2.30

Ma27 committed:
gpsd: fix build w/glibc-2.30

Ma27 committed:
policycoreutils: fix build w/glibc-2.30

Ma27 committed:
prayer: fix build w/glibc-2.30

Ma27 committed:
idutils: mark as broken

Last release was in 2012, package is unmaintained and build system is
pretty old, so we can't just replace the ancient gnulib with a newer
version without further hassle.

Ma27 committed:
libmilter: fix build w/glibc-2.30

Using RES_USE_INET6 is deprecated and has been removed by glibc 2.30[1].
Instead, there's now a function `gethostbyname2(3)` which takes the
address family as second argument to force IPv6 resolution[2].

[1] https://sourceware.org/ml/libc-alpha/2019-08/msg00029.html
[2] https://linux.die.net/man/3/gethostbyname2

Ma27 commented:
Just realized that there's glibc-2.31 released today (https://sourceware.org/ml/libc-announce/2020/msg00001.html).

I fairly doubt that this will make it into 20.03 which is why I won't push the bump onto this branch. I have some stuff locally that I'll push today and wait for the llvm changes being in staging to merge those onto this branch as well.

The biggest problem atm is texlive which is also broken on staging, so most likely not our problem. If everything goes smoothly and someone pushes freshly built bootstrap tools to our binary cache, we **may** be able to finish this before the feature freeze :)

Ma27 committed:
kermit: mark as broken

Uses internal API of the _IO_FILE struct which breaks with glibc-2.30.
Also there's no release since 2011.

Ma27 committed:
lwan: 0.2 -> 0.3, fix build w/glibc-2.30

Ma27 committed:
Merge branch 'staging' into glibc230

disassembler commented:
If this can get through staging before feature freeze, I'm fine with this going in this release. Otherwise, we'll push it to 20.09.

vcunat committed:
glibc/*.patch: revert no-op changes done in 447edaa3

I suppose the diff of the glibc-upgrade branch is a bit cleaner
without including these unnecessary changes.

vcunat commented:
_I'm sorry about my recent non-helping here._

Nitpick: it's a little weird that all edits in `pkgs/development/libraries/glibc/*.patch` files are non-meaningful (files patched in different order, different format diff headers) &ndash; but I suppose it doesn't really matter.

- - -

In any case, we need to finish tarballs.nixos.org ASAP (it would've been better weeks ago or earlier, but here we are).  I originally thought that bootstraps generated from 19.09 should be enough &ndash; that's what I tested [above](https://github.com/NixOS/nixpkgs/pull/66528#issuecomment-549169573) and they still seem to work fine (re-checked x86_64 now).

Still, I can't see a real reason to avoid generating them from recent unstable channel, especially if you think that's better for some reason (I'm not sure if gcc9 will suffice longer than gcc8).  Well, except that for the old ones I have hashes etc. ready as obtained from cache.nixos.org and thus only simple uploading remains, but that part shouldn't be hard IIRC.

By looking at the commits, we probably "can't" avoid changing all hashes from what's been currently built in the experimental Hydra jobset.  (Though if someone told me the commits they were generated from and no build impurity changed the hashes from what's in cache.nixos.org already...)

Any thoughts about this?
_After that, I'll try "forcing" someone to upload what we've agreed on._

vcunat committed:
stdenv bootstrap tools: regenerate "officialy"

From Hydra's binaries for c5aabb0 (19.09).
This time I re-tested full stdenv bootstrap on these three platforms.
https://github.com/NixOS/nixpkgs/pull/66528#issuecomment-582351637

FRidh commented on pkgs/development/tools/misc/gnum4/default.nix:
```diff
@@ -13,7 +13,14 @@ stdenv.mkDerivation {
+  patches =
+    [
+      ./s_isdir.patch
+      (fetchurl {
```
fetchpatch?

vcunat replied:
This isn't a dynamically generated patch, just simple http(s) download.

FRidh commented on pkgs/tools/archivers/sharutils/default.nix:
fetchpatch?

vcunat replied:
Same as https://github.com/NixOS/nixpkgs/pull/66528#discussion_r374265548 (so that the same discussion isn't split to two places)

Ma27 commented:
> In any case, we need to finish tarballs.nixos.org ASAP (it would've been better weeks ago or earlier, but here we are). I originally thought that bootstraps generated from 19.09 should be enough – that's what I tested above and they still seem to work fine (re-checked x86_64 now; EDIT: also aarch64 and i686, after a while).

:+1: - I talked with a few people at FOSDEM how we could improve the situation with bootstrap files (e.g. by using stuff from a binary cache by default), but this is something we should take care of after the branchoff.

> Still, I can't see a real reason to avoid generating them from recent unstable channel

Same here - I actually built the aarch64 bootstrap stuff with a gcc8 since the gcc9 changes weren't on master back then, so this should be fine :)

> By looking at the commits, we probably "can't" avoid changing all hashes from what's been currently built in the experimental Hydra jobset. (Though if someone told me the commits they were generated from and no build impurity changed the hashes from what's in cache.nixos.org already...)

Unfortunately I'm not sure anymore - I only built the aarch64 stuff as it was broken with glibc before :/

> Any thoughts about this?
After that, I'll try "forcing" someone to upload what we've agreed on.

Overall :+1:

vcunat commented:
Oh, I missed one bit: i686-linux.  I had re-tested glibc build but the next step of building the default compiler fails :-/  https://hydra.nixos.org/job/nixpkgs/glibc-230/wine.x86_64-linux

vcunat commented:
I worked around that by `enableLTO = !stdenv.isi686;` which sounds acceptable to me.  Interested parties can improve that later. I suspect update of bootstrapping tarballs would _not_ be involved in this issue anyway...

OK, let's take the easier way: c5aabb0d60 from 19.09.  This time I re-tested _full_ stdenv bootstrap on these three platforms.

@peti: I heard you have credentials for tarballs.nixos.org &ndash; can you upload these for us?

#### We need to upload:
- `/nix/store/21zzaia2nca8p5vpcrklrdqlj843q2wf-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz` -> http://tarballs.nixos.org/stdenv-linux/x86_64/c5aabb0d603e2c1ea05f5a93b3be82437f5ebf31/bootstrap-tools.tar.xz
- `/nix/store/l1r0lyzdr2n5qqz68rwxnh71qjd5s4ay-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz` -> http://tarballs.nixos.org/stdenv-linux/i686/c5aabb0d603e2c1ea05f5a93b3be82437f5ebf31/bootstrap-tools.tar.xz
- `/nix/store/fwrp466vixrzc63l4kvlslalfsxkjnvk-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz` -> http://tarballs.nixos.org/stdenv-linux/aarch64/c5aabb0d603e2c1ea05f5a93b3be82437f5ebf31/bootstrap-tools.tar.xz
- _optionally_ also relocate this one without changing &ndash; we have aarch64 among officially supported platforms, so better not fetch it from some private server (though you'll typically substitute it from binary cache anyway)
  `/nix/store/y5kymk0pdj7pszxskf9xazg5825alizg-busybox` -> http://tarballs.nixos/org/stdenv-linux/aarch64/bb3ef8a95c9659596b8a34d27881cd30ffea2f9f/busybox

#### Evidence
... that we're not switching to something suspicious.  (Bootstrapping is sensitive; see e.g. classic "Trusting Trust".)

```bash
## It is official 19.09-included commit
git checkout c5aabb0d60
git branch --contains | grep release-19.09

## Verify that /nix/store paths match
for system in x86_64-linux i686-linux aarch64-linux; do
    cat "$(nix-build pkgs/top-level/release.nix -Q -A stdenvBootstrapTools.${system}.dist)"/nix-support/hydra-build-products
done | grep tarball 2>/dev/null
# file tarball /nix/store/21zzaia2nca8p5vpcrklrdqlj843q2wf-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
# file tarball /nix/store/l1r0lyzdr2n5qqz68rwxnh71qjd5s4ay-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz
# file tarball /nix/store/fwrp466vixrzc63l4kvlslalfsxkjnvk-stdenv-bootstrap-tools/on-server/bootstrap-tools.tar.xz

## The optional re-upload without changing:
nix-build pkgs/stdenv/linux/bootstrap-files/aarch64.nix -A busybox
# copying path '/nix/store/y5kymk0pdj7pszxskf9xazg5825alizg-busybox' from 'https://cache.nixos.org'...
# /nix/store/y5kymk0pdj7pszxskf9xazg5825alizg-busybox
```

peti commented:
> @peti: I heard you have credentials for tarballs.nixos.org – can you upload these for us?

I am sorry, but that seems to be a misunderstanding. I have no privileged access to tarballs.nixos.org. I don't even know who does, I'm afraid.

vcunat commented:
Thanks for the fast reaction.  I guess I have to ping @edolstra, as I don't know about anyone else.

vcunat commented:
@Ma27

> I talked with a few people at FOSDEM how we could improve the situation with bootstrap files (e.g. by using stuff from a binary cache by default), but this is something we should take care of after the branchoff.

There seems to be some misunderstanding.  Bootstrap files *are* substituted from cache.nixos.org (if present); see e.g. log from `busybox` a couple posts above.
- The [regular Hydra job](https://hydra.nixos.org/job/nixpkgs/trunk/stdenvBootstrapTools.x86_64-linux.dist) as of now doesn't produce these as _separate_ files, so it's the same files but nested deeper in other /nix/store paths and can't be directly re-used. _Still, it's not difficult to `nix build $file` and then `nix-prefetch-url file://$file`_
- Users with different location of nix store can't use the official binary cache (the hash parts are changed), so it's good to have working fixed-output fetches for them to bootstrap from.  Hence tarballs.nixos.org
- These fetches also serve as backup in case the files got garbage-collected from the cache, though I suspect in our case that's just a theoretical possibility for now.

Maybe I forgot some point.  We certainly had at least one GitHub discussion about the (non-)utility of tarballs.nixos.org.

vcunat committed:
gcc: disable LTO on i686

as a workaround to fix build after updating stdenv bootstrap
(in the followup commit).  Interested parties can improve this later.

edolstra commented:
@vcunat Thanks, I've copied the files.

vcunat commented:
Confirmed, except for the optional one (but that one doesn't matter).

edolstra commented:
No, I copied it as well. Note that you have a typo in the URL (`http://tarballs.nixos/org...`).

lblasc commented:
@vcunat @Ma27 it was a pleasure working with you guys! I've started this as a learning experience to investigate how NixOS bootstrapping works, I'm very pleased to have contribution which rebuilds the whole dependency tree :fire: on three major archs :grin:

vcunat commented:
Oh, right, typo; pushed as cf242659fe8.  First x86_64 binaries should [start compiling today](https://hydra.nixos.org/eval/1568809); the rest when staging-next get rotated.

@lblasc: it's a bit unlucky that it was a difficult case and took us so much time &ndash; about half a year.  Usually changes aren't so cumbersome, even mass rebuilds get merged basically every week.

vcunat commented:
Repeated note: 32-bit ARMs are quite certainly broken by this PR.  /cc @dezgeg (again)  Regenerating bootstrap tools for them might suffice to fix them, but I don't have a good way of testing the builds natively.

I suspect cross-building these platforms might keep working, except I _now_ see that this PR broke evaluation of (all?) cross stuff :face_with_head_bandage:

Ma27 commented:
> I suspect cross-building these platforms might keep working, except I now see that this PR broke evaluation of (all?) cross stuff

Can you please point me to a Hydra eval where I can see the errors?

jcumming commented:
Uh-oh. These bootstrap tools no work on SLES 11-SP4 `3.0.101-108.87-default` kernels:

```
Unpacking the bootstrap tools...
Patching the bootstrap tools...
FATAL: kernel too old
```

Ma27 commented:
Just to properly understand your issue: you're using Nix on Sles11sp4 and our new bootstrap tools break the setup, right?

However I'd like to point out that I'm pretty much against putting additional effort in fixing stuff for [unmaintained versions](https://www.suse.com/lifecycle/) of other distros (well unless you're paying for LTSS, but that's another story), but that's just my opinion :)

vcunat commented:
@jcumming: did it even work before this PR?  I've never looked into this for SLES/SLED.  We haven't changed this part for quite a long time:
> "--enable-kernel=3.2.0" # can't get below with glibc >= 2.26

We have special patch to allow RHEL 6 -like kernels.  That reminds me I last verified the real compatibility with that on glibc-2.26-131... so even though you don't get this nice refusal, some parts might not work on those kernels anymore.  Well, today I suspect we might even drop RHEL 6 support :woman_shrugging: I personally don't feel it's important anymore.

vcunat commented:
To be clear, it isn't that much about bootstap... but about running anything built with our stdenv.

vcunat commented:
Actually, I think previous service packs with 2.6.32 kernel would give you the error message by co-incidence with RHEL 6 :-)
